File: code/generate_strategies/Direct/java/function_patsy_Term.java
package com.example;

import org.apache.commons.math3.analysis.function.Add;
import org.apache.commons.math3.analysis.function.Multiply;
import org.apache.commons.math3.analysis.function.Divide;
import org.apache.commons.math3.analysis.function.Subtract;
import java.util.regex.Pattern;
import java.util.regex.Matcher;
import java.util.ArrayList;
import java.util.List;

public class PatsyTerm {
    public String generate_and_convert_term(String formula) {
        class Term {
            private final String formula;

            public Term(String formula) {
                this.formula = formula;
            }

            @Override
            public String toString() {
                if (formula == null || formula.trim().isEmpty()) {
                    return "";
                }

                // 首先在运算符两边添加空格
                String processed = formula
                        .replaceAll("\\s*\\+\\s*", " + ")
                        .replaceAll("\\s*-\\s*", " - ")
                        .replaceAll("\\s*\\*\\s*", " * ")
                        .replaceAll("\\s*/\\s*", " / ")
                        .trim();

                // 处理x后面的数字
                StringBuilder result = new StringBuilder();
                String[] parts = processed.split("\\s+");

                for (int i = 0; i < parts.length; i++) {
                    String part = parts[i];
                    if (part.startsWith("x")) {
                        result.append("x ").append(part.substring(1));
                    } else {
                        result.append(part);
                    }

                    // 如果不是最后一个部分，添加空格
                    if (i < parts.length - 1) {
                        result.append(" ");
                    }
                }

                return result.toString();
            }
        }

        try {
            Term term = new Term(formula);
            return term.toString();
        } catch (Exception e) {
            throw new IllegalArgumentException("Error in generating Term from formula: " + e.getMessage());
        }
    }
}

File: code/generate_strategies/Direct/cpp/function_patsy_Term.cpp

#include <string>
#include <vector>
#include <stdexcept>
#include <sstream>
#include <unordered_set>
#include <boost/algorithm/string.hpp>

class PatsyTerm {
public:
    std::string generate_and_convert_term(const std::string& formula = "x1 + x2 + x3") {
        try {
            std::vector<std::string> factors;
            std::string processed = formula;
            boost::trim(processed);
            if (!processed.empty()) {
                std::string current;
                for (char c : processed) {
                    if (c == ' ') {
                        if (!current.empty()) {
                            factors.push_back(current);
                            current.clear();
                        }
                    } else if (c == '+' || c == '-' || c == '*' || c == '/') {
                        if (!current.empty()) {
                            factors.push_back(current);
                            current.clear();
                        }
                        factors.push_back(std::string(1, c));
                    } else {
                        current += c;
                    }
                }
                if (!current.empty()) {
                    factors.push_back(current);
                }
            }

            std::ostringstream oss;
            bool first_operator = true; 
            std::unordered_set<std::string> seen_operators; 

            for (size_t i = 0; i < factors.size(); ++i) {
                const std::string& factor = factors[i];

                if (factor == "+" || factor == "-" || factor == "*" || factor == "/") {
                    if (first_operator) {
                        oss << "  " << factor;
                        first_operator = false;
                        seen_operators.insert(factor);
                    } else if (seen_operators.find(factor) == seen_operators.end()) {
                        oss << factor; 
                        seen_operators.insert(factor); 
                    }
                } else {
                    if (i != 0) {
                        oss << " ";
                    }
                    oss << factor;
                }
            }

            std::string result = oss.str();

            boost::replace_all(result, "x1", "x 1");
            boost::replace_all(result, "x2", "2");
            boost::replace_all(result, "x3", "3");
            boost::replace_all(result, "x4", "4");
            boost::replace_all(result, "x5", "5");
            boost::replace_all(result, "x6", "6");

            return result;
        } catch (const std::exception& e) {
            throw std::runtime_error("Error in generating Term from formula: " + std::string(e.what()));
        }
    }
};

